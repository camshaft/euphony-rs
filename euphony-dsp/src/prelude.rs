pub use crate::sample::{
    DefaultRate as Rate, DefaultSample as Sample, Rate as RateTrait, Sample as SampleTrait,
};
pub use euphony_node::{Buffer, Input, Node};
pub use euphony_units::zip::Zip;
