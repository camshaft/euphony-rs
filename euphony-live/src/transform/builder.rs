use super::*;
use std::cell::RefCell;

thread_local! {
    static INSTRUCTIONS: RefCell<Builder> = RefCell::new(Builder::default());
}

fn borrow<F: FnOnce(&mut Builder) -> R, R>(f: F) -> R {
    INSTRUCTIONS.with(|b| f(&mut b.borrow_mut()))
}

#[derive(Clone, Debug, Default)]
struct Builder {
    once_instructions: Vec<Transform>,
    instructions: Vec<Transform>,
    in_once: bool,
    skip_stack: Vec<usize>,
    local: usize,
    cells: Vec<Event>,
    outputs: Vec<(String, Output)>,
    inputs: Vec<(String, Input)>,
    constants: HashMap<Event, Local>,
    tables: usize,
}

impl Builder {
    fn build(self) -> Config {
        let locals = self.local;
        let instructions = self.instructions;
        let databases = self.tables;
        let cells = self.cells;
        let outputs = self.outputs;
        let inputs = self.inputs;
        let once_instructions = self.once_instructions;

        Config {
            locals,
            instructions,
            once_instructions,
            cells,
            databases,
            outputs,
            inputs,
        }
    }

    fn swap_once(&mut self) {
        self.in_once = !self.in_once;
    }
}

fn emit<T: Into<Transform>>(transform: T) -> usize {
    borrow(|b| {
        let t = transform.into();

        if matches!(t, Transform::Constant { .. }) || b.in_once {
            b.once_instructions.push(t);
            return 0;
        }

        let idx = b.instructions.len();
        b.instructions.push(t);
        for s in &mut b.skip_stack {
            *s += 1;
        }
        idx
    })
}

pub fn constant<T: Into<Event>>(value: T) -> Local {
    let value = value.into();
    if let Some(l) = borrow(|b| b.constants.get(&value).copied()) {
        return l;
    }

    let destination = local();

    borrow(|b| b.constants.insert(value.clone(), destination));

    emit(Transform::Constant { value, destination });
    destination
}

macro_rules! ext {
    ($name:ident, $variant:ident, $ty:ty, $field:ident) => {
        pub fn $name<T: Into<$ty>>($field: T) -> Local {
            let l = local();
            emit(Transform::$variant {
                $field: $field.into(),
                destination: l,
            });
            l
        }
    };
}

ext!(math, Math, Math, math);
ext!(logic, Logic, Logic, logic);
ext!(midi, Midi, Midi, midi);

pub fn local() -> Local {
    borrow(|state| {
        let id = state.local;
        state.local += 1;
        Local { id }
    })
}

pub fn argument(index: usize) -> Local {
    let destination = local();

    emit(Transform::Argument { index, destination });

    destination
}

pub fn table() -> TableBuilder {
    borrow(|state| {
        let id = state.tables;
        state.tables += 1;
        TableBuilder { id }
    })
}

pub struct TableBuilder {
    id: usize,
}

impl TableBuilder {
    pub fn push<G: Into<Vec<Local>>>(&self, groups: G, value: Local) {
        emit(Transform::Table {
            id: self.id,
            table: Table::Push {
                source: value,
                groups: groups.into(),
            },
        });
    }

    pub fn group_len(&self, group: Local) -> Local {
        let destination = local();

        emit(Transform::Table {
            id: self.id,
            table: Table::GroupLen { group, destination },
        });

        destination
    }

    pub fn select_index(&self, group: Local, index: Local) -> Local {
        let destination = local();

        emit(Transform::Table {
            id: self.id,
            table: Table::SelectIndex {
                group,
                index,
                destination,
            },
        });

        destination
    }

    pub fn send_all<G: Into<Vec<Local>>>(&self, groups: G, output: Output) {
        emit(Transform::Table {
            id: self.id,
            table: Table::SendAll {
                groups: groups.into(),
                output: constant(output.id as i64),
            },
        });
    }
}

pub fn cell<I: Into<Event>>(init: I) -> Cell {
    borrow(|state| {
        let id = state.cells.len();
        state.cells.push(init.into());
        Cell { id }
    })
}

impl Cell {
    pub fn load(&self) -> Local {
        let l = local();
        emit(Transform::Load {
            cell: *self,
            destination: l,
        });
        l
    }

    pub fn store(&self, value: Local) {
        value.store(*self)
    }
}

pub fn output<N: Into<String>>(name: N) -> Output {
    borrow(|state| {
        let id = state.outputs.len();
        let v = Output { id, bus: None };
        state.outputs.push((name.into(), v));
        v
    })
}

pub fn input<N: Into<String>>(name: N) -> Input {
    borrow(|state| {
        let id = state.inputs.len();
        let v = Input { id, is_bus: false };
        state.inputs.push((name.into(), v));
        v
    })
}

pub fn bus() -> (Input, Output) {
    let i = input("");
    let o = output("");

    // remove the physical I/O
    borrow(|state| {
        state.inputs.last_mut().unwrap().1.is_bus = true;
        state.outputs.last_mut().unwrap().1.bus = Some(i.id);
    });

    (i, o)
}

pub fn cond<F: FnOnce() -> R, R>(condition: Local, f: F) -> R {
    // push the condition instruction
    let idx = emit(Transform::Condition {
        skip_count: 0,
        condition: condition.id,
    });

    // push a condition on to the stack
    borrow(|v| v.skip_stack.push(0));

    // execute the truthy condition
    let r = f();

    // get the final skip count
    borrow(|b| {
        let skip_count = b.skip_stack.pop().unwrap();
        b.instructions[idx] = Transform::Condition {
            skip_count,
            condition: condition.id,
        };
    });

    // return the closure value
    r
}

pub fn build<F: FnOnce()>(f: F) -> Config {
    f();
    borrow(|b| core::mem::take(b).build())
}

pub fn once<F: FnOnce() -> R, R>(f: F) -> R {
    borrow(|b| b.swap_once());
    let r = f();
    borrow(|b| b.swap_once());
    r
}

impl Local {
    pub fn assign_to(&self, destination: Local) {
        emit(Transform::Move {
            source: *self,
            destination,
        });
    }

    pub fn filter<T: Into<Filter>>(&self, filter: T) -> Local {
        let l = local();
        emit(Transform::Filter {
            source: *self,
            filter: filter.into(),
            destination: l,
        });
        l
    }

    pub fn select<T: Into<Select>>(&self, select: T) -> Local {
        let l = local();
        emit(Transform::Select {
            source: *self,
            select: select.into(),
            destination: l,
        });
        l
    }

    pub fn theory<T: Into<Theory>>(&self, theory: T) -> Local {
        let l = local();
        emit(Transform::Theory {
            source: *self,
            theory: theory.into(),
            destination: l,
        });
        l
    }

    pub fn inspect<S: Into<String>>(&self, label: S) -> Self {
        emit(Transform::Inspect {
            label: label.into(),
            source: *self,
        });
        *self
    }

    pub fn store(&self, cell: Cell) {
        emit(Transform::Store {
            source: *self,
            cell,
        });
    }

    pub fn not(&self) -> Local {
        logic(Logic::Not(vec![*self]))
    }

    pub fn or(&self, rhs: Local) -> Local {
        logic(Logic::Any(vec![*self, rhs]))
    }

    pub fn eq(&self, rhs: Local) -> Local {
        logic(Logic::Equal(vec![*self, rhs]))
    }

    pub fn add(&self, rhs: Local) -> Local {
        math(Math::Add(vec![*self, rhs]))
    }

    pub fn sub(&self, rhs: Local) -> Local {
        math(Math::Sub(vec![*self, rhs]))
    }

    pub fn div(&self, rhs: Local) -> Local {
        math(Math::Div(vec![*self, rhs]))
    }

    pub fn rem(&self, rhs: Local) -> Local {
        math(Math::Mod(vec![*self, rhs]))
    }

    // TODO
    // pub fn load()
    // pub fn store()
}

#[derive(Clone, Copy, Debug)]
pub struct Output {
    pub(super) id: usize,
    pub(super) bus: Option<usize>,
}

impl Output {
    pub fn send(&self, local: Local) {
        emit(Transform::Send {
            source: local,
            output: constant(self.id as i64),
        });
    }

    pub fn send_after(&self, local: Local, delay: Local) {
        emit(Transform::SendAfter {
            source: local,
            delay,
            output: constant(self.id as i64),
        });
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Input {
    pub(super) id: usize,
    pub(super) is_bus: bool,
}

impl Input {
    pub fn is_from(&self) -> Local {
        argument(1).eq(constant(self.id as i64))
    }
}
