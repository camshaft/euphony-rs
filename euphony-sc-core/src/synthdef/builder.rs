#![allow(dead_code)]

use crate::{osc::control, param::Param, synthdef::CalculationRate};
use core::{fmt, marker::PhantomData, ops};

#[derive(Clone, Copy)]
pub struct Value(u32, u32);

impl Value {
    pub fn calculation_rate(&self) -> CalculationRate {
        self.as_variant().calculation_rate()
    }

    pub fn from_f32(value: f32) -> Self {
        Variant::Const(value).as_value()
    }

    pub fn from_i32(value: i32) -> Self {
        Self::from_f32(value as _)
    }

    pub fn from_parameter(value: Parameter) -> Self {
        Variant::Parameter(value).as_value()
    }

    pub fn from_parameter_id(value: u32) -> Self {
        Variant::Parameter(Parameter(value)).as_value()
    }

    pub fn from_ugen(value: Ugen) -> Self {
        Variant::Ugen(value).as_value()
    }

    pub fn as_variant(self) -> Variant {
        match self.0 {
            v if v == u32::MAX - 1 => Variant::Parameter(Parameter(self.1)),
            u32::MAX => Variant::Const(f32::from_bits(self.1)),
            _ => Variant::Ugen(Ugen(self.0, self.1)),
        }
    }

    pub fn as_osc(self) -> Option<control::Value> {
        match self.as_variant() {
            Variant::Const(value) => Some(control::Value::Float(value)),
            _ => None,
        }
    }
}

impl fmt::Debug for Value {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.as_variant().fmt(f)
    }
}

impl From<i32> for Value {
    fn from(v: i32) -> Self {
        Value::from_i32(v)
    }
}

impl From<f32> for Value {
    fn from(v: f32) -> Self {
        Value::from_f32(v)
    }
}

impl From<Parameter> for Value {
    fn from(param: Parameter) -> Self {
        Value::from_parameter(param)
    }
}

impl From<Ugen> for Value {
    fn from(ugen: Ugen) -> Self {
        Value::from_ugen(ugen)
    }
}

impl From<Value> for Variant {
    fn from(value: Value) -> Self {
        value.as_variant()
    }
}

impl From<Variant> for Value {
    fn from(value: Variant) -> Self {
        value.as_value()
    }
}

impl From<control::Value> for Value {
    fn from(value: control::Value) -> Self {
        match value {
            control::Value::Int(v) => v.into(),
            control::Value::Float(v) => v.into(),
        }
    }
}

impl<T> ops::Add<T> for Value
where
    T: Into<Value>,
{
    type Output = Value;

    fn add(self, _rhs: T) -> Self::Output {
        // let rhs = rhs.into();
        todo!()
    }
}

#[derive(Clone, Copy, Debug)]
pub enum Variant {
    Const(f32),
    Parameter(Parameter),
    Ugen(Ugen),
}

impl Variant {
    pub const fn calculation_rate(&self) -> CalculationRate {
        match self {
            Self::Const(_) => CalculationRate::Scalar,
            Self::Parameter(v) => v.calculation_rate(),
            Self::Ugen(v) => v.calculation_rate(),
        }
    }

    pub fn as_value(self) -> Value {
        match self {
            Self::Const(v) => Value(u32::MAX, v.to_bits()),
            Self::Parameter(v) => Value(u32::MAX - 1, v.0),
            Self::Ugen(v) => Value(v.0, v.1),
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Parameter(u32);

impl Parameter {
    pub const fn new(idx: u32) -> Self {
        Self(idx)
    }

    pub const fn calculation_rate(self) -> CalculationRate {
        CalculationRate::Control
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Ugen(u32, u32);

impl Ugen {
    pub const fn calculation_rate(self) -> CalculationRate {
        match self.0 >> 30 {
            0b00 => CalculationRate::Scalar,
            0b01 => CalculationRate::Control,
            0b10 => CalculationRate::Audio,
            _ => CalculationRate::Demand,
        }
    }
}

pub trait Parameters: 'static + Sized {
    type Desc;

    fn new<F>(f: F) -> Self
    where
        F: FnOnce(fn() -> Self::Desc) -> SynthDescRef;
}

#[derive(Clone)]
pub struct Synth {
    id: crate::osc::node::Id,
    track: crate::track::Handle,
    synthdef: &'static str,
}

impl Synth {
    pub fn new(
        id: crate::osc::node::Id,
        track: crate::track::Handle,
        synthdef: &'static str,
    ) -> Self {
        Self {
            id,
            track,
            synthdef,
        }
    }

    pub fn name(&self) -> &'static str {
        self.synthdef
    }

    pub fn id(&self) -> crate::osc::node::Id {
        self.id
    }

    pub fn track_name(&self) -> &str {
        use crate::track::Track;
        self.track.name()
    }
}

impl fmt::Debug for Synth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Synth")
            .field("id", &self.id)
            .field("track", &self.track_name())
            .field("synthdef", &self.synthdef)
            .finish()
    }
}

impl Drop for Synth {
    fn drop(&mut self) {
        use crate::track::Track;
        self.track.free(self.id)
    }
}

pub struct SynthDesc {
    name: String,
    desc: Vec<u8>,
}

impl SynthDesc {
    pub fn as_ref(&'static self) -> SynthDescRef {
        SynthDescRef {
            name: &self.name,
            desc: &self.desc,
        }
    }
}

impl fmt::Debug for SynthDesc {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("SynthDesc")
            .field("name", &self.name)
            .finish()
    }
}

#[derive(Clone, Copy)]
pub struct SynthDescRef {
    name: &'static str,
    desc: &'static [u8],
}

impl fmt::Debug for SynthDescRef {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("SynthDesc")
            .field("name", &self.name)
            .finish()
    }
}

pub struct SynthDef<Params> {
    desc: SynthDescRef,
    params: PhantomData<Params>,
}

impl<Params> SynthDef<Params> {
    pub fn new(desc: SynthDescRef) -> Self {
        Self {
            desc,
            params: PhantomData,
        }
    }

    pub fn name(&self) -> &'static str {
        &self.desc.name
    }

    pub fn desc(&self) -> &'static [u8] {
        &self.desc.desc
    }
}

impl<Params> fmt::Debug for SynthDef<Params> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("SynthDef")
            .field("name", &self.name())
            .finish()
    }
}

use std::cell::RefCell;

struct UgenSpec {
    name: &'static str,
    rate: CalculationRate,
    inputs: Vec<Vec<Value>>,
    outputs: Vec<CalculationRate>,
    special_index: i16,
}

#[derive(Default)]
struct Context {
    params: Vec<(&'static str, f32)>,
    ugens: Vec<UgenSpec>,
}

impl Context {
    fn build(&mut self, name: &'static str) -> SynthDesc {
        // TODO
        SynthDesc {
            name: name.to_owned(),
            desc: vec![],
        }
    }

    fn param(&mut self, id: u32, name: &'static str, default: f32) -> Param {
        let idx = id as usize;
        if idx >= self.params.len() {
            self.params.resize(idx + 1, ("", 0.0));
        }
        self.params[idx] = (name, default);
        param_instance(id)
    }

    fn ugen(&mut self, ugen: UgenSpec) -> Ugen {
        let id = self.ugens.len() as _;
        let idx = 0;
        self.ugens.push(ugen);
        Ugen(idx, id)
    }
}

thread_local! {
    static CONTEXT: RefCell<Context> = RefCell::new(Context::default());
}

pub fn synthdef<F: FnOnce()>(name: &'static str, b: F) -> SynthDesc {
    b();
    CONTEXT.with(|c| c.take()).build(name)
}

pub fn external_synthdef(name: &'static str, desc: &'static [u8]) -> SynthDescRef {
    SynthDescRef { name, desc }
}

pub fn param(id: u32, name: &'static str, default: f32) -> Param {
    CONTEXT.with(|c| c.borrow_mut().param(id, name, default))
}

pub fn param_instance(id: u32) -> Param {
    let v: Value = Parameter::new(id).into();
    v.into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn value_size_test() {
        assert_eq!(core::mem::size_of::<Value>(), 8);
    }
}
