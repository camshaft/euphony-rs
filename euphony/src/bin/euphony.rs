fn main() {
    euphony_cli::main()
}
