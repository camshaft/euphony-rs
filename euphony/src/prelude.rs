pub use crate::{
    buffer::{self, Buffer, BufferExt},
    cell::Cell,
    ext::*,
    group::*,
    processor::Processor,
    processors::{
        ext::*,
        input::{self, *},
        *,
    },
    rand,
    runtime::{primary, spawn},
    section::section,
    time::{now, set_tempo, tempo},
    units::{
        pitch::{
            frequency::*,
            mode::{self, Mode},
            tuning, Interval,
        },
        time::{Beat, Tempo},
        zip::Zip as ZipExt,
    },
};
pub use euphony_samples as samples;
pub use futures::{FutureExt, StreamExt};

#[macro_export]
macro_rules! delay {
    ($n:literal) => {
        $crate::prelude::delay!($crate::prelude::Beat($n, 1))
    };
    ($n:literal / $d:literal) => {
        $crate::prelude::delay!($crate::prelude::Beat($n, $d))
    };
    ($expr:expr) => {
        $crate::prelude::DelayExt::delay($expr).await
    };
}
pub use delay;

pub mod western {
    pub use super::{mode::western::*, tuning::western::*};
}
