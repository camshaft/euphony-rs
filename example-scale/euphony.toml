[synths]
chiplead = "gh://sonic-pi-net/sonic-pi/etc/synthdefs/compiled/sonic-pi-chiplead.scsyndef"
bass = "gh://sonic-pi-net/sonic-pi/etc/synthdefs/compiled/sonic-pi-blade.scsyndef"

[samples]
bd = "clean://tidalcycles/sounds-tr808-fischer/bd8"
sd = "clean://tidalcycles/sounds-tr808-fischer/sd8"
hh = "clean://tidalcycles/sounds-tr808-fischer/ch8"
cy = "clean://tidalcycles/sounds-tr808-fischer/cy8"
ht = "clean://tidalcycles/sounds-tr808-fischer/ht8"
lt = "clean://tidalcycles/sounds-tr808-fischer/lt8"
mt = "clean://tidalcycles/sounds-tr808-fischer/mt8"
