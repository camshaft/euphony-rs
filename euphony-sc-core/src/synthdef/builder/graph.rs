use super::{
    value::{Output, Value, ValueVec, Variant},
    UgenSpec,
};
use crate::synthdef::{self, CalculationRate};
use std::collections::{hash_map::Entry, HashMap};

#[derive(Debug)]
struct Node {
    spec: UgenSpec,
    inputs: Vec<ValueVec>,
}
type Input = (u32, u32);

#[derive(Debug)]
pub struct Graph {
    nodes: Vec<Node>,
}

impl Default for Graph {
    fn default() -> Self {
        let node = Node {
            spec: UgenSpec {
                name: "Control",
                special_index: 0,
                rate: |_| CalculationRate::Control,
                outputs: 0,
                output_rate: |_, _| CalculationRate::Control,
            },
            inputs: vec![],
        };
        Self { nodes: vec![node] }
    }
}

impl Graph {
    pub fn insert(&mut self, spec: UgenSpec) -> Ugen {
        let id = self.nodes.len() as u32;

        let node = Node {
            spec,
            inputs: vec![],
        };
        self.nodes.push(node);

        Ugen {
            node: &mut self.nodes[id as usize],
            ugen: id,
        }
    }

    pub fn build(mut self, params: usize) -> (Vec<f32>, Vec<synthdef::UGen>) {
        self.nodes[0].spec.outputs = params;

        let mut consts = ConstIds::default();

        // TODO multichannel expansion

        for node in &self.nodes {
            for input in &node.inputs {
                for input in input.iter() {
                    let _input: synthdef::Input = match input.as_variant() {
                        Variant::Const(v) => consts.get(v),
                        Variant::Output(v) => v.into(),
                        Variant::Parameter(v) => v.into(),
                    };
                }
            }
        }

        let consts = consts.values;
        let ugens = vec![];
        (consts, ugens)
    }
}

#[derive(Default)]
struct ConstIds {
    values: Vec<f32>,
    ids: HashMap<u32, i32>,
}

impl ConstIds {
    pub fn get(&mut self, value: f32) -> synthdef::Input {
        match self.ids.entry(value.to_bits()) {
            Entry::Vacant(entry) => {
                let index = self.values.len() as i32;
                self.values.push(value);
                entry.insert(index);
                synthdef::Input::Constant { index }
            }
            Entry::Occupied(entry) => synthdef::Input::Constant {
                index: *entry.get(),
            },
        }
    }
}

pub struct Ugen<'a> {
    node: &'a mut Node,
    ugen: u32,
}

impl<'a> Ugen<'a> {
    pub fn input(&mut self, value: impl Into<ValueVec>) {
        self.node.inputs.push(value.into());
    }

    pub fn finish<T: OutputSpec>(self) -> T {
        let ugen = self.ugen;
        let count = self.node.spec.outputs as u32;
        let mut builder = OutputBuilder {
            ugen,
            output: 0,
            count,
        };
        let outputs = T::output(&mut builder);
        debug_assert_eq!(builder.output, count);
        outputs
    }
}

pub struct OutputBuilder {
    ugen: u32,
    output: u32,
    count: u32,
}

impl OutputBuilder {
    pub fn count(&self) -> usize {
        self.count as usize
    }

    pub fn output(&mut self) -> Value {
        let id = self.output;
        self.output += 1;
        Output::new(self.ugen, id).into()
    }
}

pub trait OutputSpec {
    fn output(builder: &mut OutputBuilder) -> Self;
}

impl OutputSpec for Value {
    fn output(builder: &mut OutputBuilder) -> Self {
        builder.output()
    }
}

impl OutputSpec for ValueVec {
    fn output(builder: &mut OutputBuilder) -> Self {
        let count = builder.count();
        (0..count).map(|_| builder.output()).collect()
    }
}

impl OutputSpec for (Value, Value) {
    fn output(builder: &mut OutputBuilder) -> Self {
        let a = builder.output();
        let b = builder.output();
        (a, b)
    }
}

impl OutputSpec for (Value, Value, Value) {
    fn output(builder: &mut OutputBuilder) -> Self {
        let a = builder.output();
        let b = builder.output();
        let c = builder.output();
        (a, b, c)
    }
}

impl OutputSpec for (Value, Value, Value, Value) {
    fn output(builder: &mut OutputBuilder) -> Self {
        let a = builder.output();
        let b = builder.output();
        let c = builder.output();
        let d = builder.output();
        (a, b, c, d)
    }
}
