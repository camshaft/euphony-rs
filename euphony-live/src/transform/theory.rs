use euphony::prelude::{Beat, Tempo};

use super::*;

#[derive(Clone, Debug)]
pub enum Theory {
    FromMidi,
    ToMidi,
    Collapse {
        mode: Local,
        // TODO configure strategy
    },
    Expand {
        mode: Local,
        // TODO configure strategy
    },
    BeatDuration {
        tempo: Local,
    },
}

impl Theory {
    pub fn apply(&self, event: &Event, locals: &[Event]) -> Option<Event> {
        match self {
            Self::FromMidi => {
                let n = event.as_number()?;
                Some(((n - 69) / 12).into())
            }
            Self::ToMidi => {
                let i = event.as_number()?;
                let i = i * 12 + 69;
                Some(i.into())
            }
            Self::Collapse { mode } => {
                let i = event.as_number()?;
                let m = locals[mode.id].as_mode()?;
                let i = m.checked_collapse(i.into(), Default::default())?;
                Some(i.into())
            }
            Self::Expand { mode } => {
                let i = event.as_number()?;
                let m = locals[mode.id].as_mode()?;
                let i = m.checked_expand(i.into(), Default::default())?;
                Some(i.into())
            }
            Self::BeatDuration { tempo } => {
                let beats = event.as_number()?;
                let beats = Beat::from(beats);
                let tempo = locals[tempo.id].as_number()?;
                let tempo = Tempo::from(tempo);

                let duration = tempo * beats;

                Some(duration.into())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use builder::*;
    use euphony::prelude::{western::*, *};
    use insta::assert_debug_snapshot;

    #[test]
    fn collapse_expand() {
        let mut out = 0;

        let config = build(|| {
            let key = local();
            let source_mode = local();
            let target_mode = local();

            let collapsed = key.theory(Theory::Collapse { mode: source_mode });
            let expand = collapsed.theory(Theory::Expand { mode: target_mode });

            out = expand.id;
        });

        let mut tests = vec![];

        for interval in 0..12 {
            tests.push((Interval(interval, 12).reduce(), MAJOR, MINOR));
        }

        let mut state = config.init_offline();

        let mut outputs = vec![];

        for (key, source_mode, target_mode) in tests {
            state[0] = key.into();
            state[1] = source_mode.into();
            state[2] = target_mode.into();

            state.apply(&[]);

            outputs.push((state[0].clone(), state[out].clone()));
        }

        assert_debug_snapshot!(outputs);
    }
}
