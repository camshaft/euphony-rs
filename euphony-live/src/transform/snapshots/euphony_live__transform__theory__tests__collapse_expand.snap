---
source: euphony-live/src/transform/theory.rs
assertion_line: 102
expression: outputs
---
[
    (
        Internal(
            Number(
                0,
            ),
        ),
        Internal(
            Number(
                0,
            ),
        ),
    ),
    (
        Internal(
            Number(
                1/12,
            ),
        ),
        Internal(
            Number(
                0,
            ),
        ),
    ),
    (
        Internal(
            Number(
                1/6,
            ),
        ),
        Internal(
            Number(
                1/6,
            ),
        ),
    ),
    (
        Internal(
            Number(
                1/4,
            ),
        ),
        Internal(
            Number(
                1/6,
            ),
        ),
    ),
    (
        Internal(
            Number(
                1/3,
            ),
        ),
        Internal(
            Number(
                1/4,
            ),
        ),
    ),
    (
        Internal(
            Number(
                5/12,
            ),
        ),
        Internal(
            Number(
                5/12,
            ),
        ),
    ),
    (
        Internal(
            Number(
                1/2,
            ),
        ),
        Internal(
            Number(
                5/12,
            ),
        ),
    ),
    (
        Internal(
            Number(
                7/12,
            ),
        ),
        Internal(
            Number(
                7/12,
            ),
        ),
    ),
    (
        Internal(
            Number(
                2/3,
            ),
        ),
        Internal(
            Number(
                7/12,
            ),
        ),
    ),
    (
        Internal(
            Number(
                3/4,
            ),
        ),
        Internal(
            Number(
                2/3,
            ),
        ),
    ),
    (
        Internal(
            Number(
                5/6,
            ),
        ),
        Internal(
            Number(
                2/3,
            ),
        ),
    ),
    (
        Internal(
            Number(
                11/12,
            ),
        ),
        Internal(
            Number(
                5/6,
            ),
        ),
    ),
]
